### PR TITLE
#218 Make java11 and java17 profiles activated automatically

### PR DESCRIPTION
**Pull request by jkolobok**

### CODE REVIEW

File: root-parent-pom/pom.xml [hunk 1]
```diff
@@ -377,6 +377,9 @@
         </profile>
         <profile>
             <id>java11</id>
+            <activation>
+                <jdk>11</jdk>
```
Comment (minborg) [inline root-parent-pom/pom.xml]: I am not sure but should we not use something like "[11," instead?

File: root-parent-pom/pom.xml [hunk 2]
```diff
@@ -393,6 +396,9 @@
         </profile>
         <profile>
             <id>java17</id>
+            <activation>
+                <jdk>17</jdk>
```
Comment (minborg) [inline root-parent-pom/pom.xml]: Ditto for this: "[17,"?

File: root-parent-pom/pom.xml (hunk 1 above)
Comment (JerryShea) [reply root-parent-pom/pom.xml]: agree

Comment (nicktindall) [reply root-parent-pom/pom.xml]: `[11,` would match the existing Java 17 build config (the javaXX profiles seem to be cumulative)
Screenshot from 2022-03-02 10-32-29

**Review by nicktindall — APPROVED** [review]

File: root-parent-pom/pom.xml [hunk 3]
```diff
@@ -406,7 +406,6 @@
                         <artifactId>maven-surefire-plugin</artifactId>
                         <configuration>
                             <argLine>
-                                --add-exports java.base/jdk.internal.ref=ALL-UNNAMED --add-exports=java.base/sun.nio.ch=ALL-UNNAMED
```
Comment (JerryShea) [inline root-parent-pom/pom.xml]: why remove?

Comment (minborg) [reply root-parent-pom/pom.xml]: I am also curious why..,

**Review by minborg — APPROVED** [review]

File: root-parent-pom/pom.xml (hunk 3 above)
Comment (jkolobok) [reply root-parent-pom/pom.xml]: because it exists in 11?

Comment (JerryShea) [reply root-parent-pom/pom.xml]: That sounds fair. If it works then I approve!

**Review by JerryShea — APPROVED** [review]

File: root-parent-pom/pom.xml (hunk 3 above)
Comment (nicktindall) [reply root-parent-pom/pom.xml]: I don't think it works that way....

`mvn help:effective-pom -Pjava11,java17` results in
```
<plugin>
        <artifactId>maven-surefire-plugin</artifactId>
        <version>3.0.0-M4</version>
        <configuration>
          <argLine>--add-exports=jdk.unsupported/sun.misc=ALL-UNNAMED --add-exports=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED
                                --add-opens=jdk.compiler/com.sun.tools.javac=ALL-UNNAMED --add-exports=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED
                                --add-opens=java.base/java.lang=ALL-UNNAMED --add-opens=java.base/java.lang.reflect=ALL-UNNAMED
                                --add-opens=java.base/java.io=ALL-UNNAMED --add-opens=java.base/java.util=ALL-UNNAMED</argLine>
        </configuration>
      </plugin>
```
while
`mvn help:effective-pom -Pjava11` results in
```
<plugin>
        <artifactId>maven-surefire-plugin</artifactId>
        <version>3.0.0-M4</version>
        <configuration>
          <argLine>--add-exports java.base/jdk.internal.ref=ALL-UNNAMED --add-exports
                                java.base/jdk.internal.misc=ALL-UNNAMED --add-exports java.base/sun.nio.ch=ALL-UNNAMED</argLine>
        </configuration>
      </plugin>
```
and `mvn help:effective-pom -Pjava17` results in
```
<plugin>
        <artifactId>maven-surefire-plugin</artifactId>
        <version>3.0.0-M4</version>
        <configuration>
          <argLine>--add-exports=jdk.unsupported/sun.misc=ALL-UNNAMED --add-exports=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED
                                --add-opens=jdk.compiler/com.sun.tools.javac=ALL-UNNAMED --add-exports=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED
                                --add-opens=java.base/java.lang=ALL-UNNAMED --add-opens=java.base/java.lang.reflect=ALL-UNNAMED
                                --add-opens=java.base/java.io=ALL-UNNAMED --add-opens=java.base/java.util=ALL-UNNAMED</argLine>
        </configuration>
      </plugin>
```

You may be able to make it work that way with `combine.children=append`, but perhaps it's better just to duplicate it incase the profiles are explicitly specified and the caller doesn't realise they need to enable both?

Comment (nicktindall) [reply root-parent-pom/pom.xml]: See https://blog.sonatype.com/2011/01/maven-how-to-merging-plugin-configuration-in-complex-projects/

Comment (nicktindall) [reply root-parent-pom/pom.xml]: That's 11 years old though... so I wonder if it's still correct, Maven moves pretty slowly

Comment (JerryShea) [reply root-parent-pom/pom.xml]: The caller should not need to enable any more

Comment (nicktindall) [reply root-parent-pom/pom.xml]: But you can explicitly set which profiles are enabled I think... which would override the JDK detection.

Comment (jkolobok) [reply root-parent-pom/pom.xml]: Ok. I'll restore the old command line and change profile JDKs to [11,16] and [17,

**Pull request merged**